### ISSUE TITLE
Performance issue when using an eqx.Module class as an argument

### ISSUE THREAD

**Issue by ToshiyukiBandai** [patrick-kidger/equinox#576]
Hi Equinox community,

I want to use an eqx.Module class to store variables, but it seems like the performance is worse than directly using the variables as an argument. Please see the example below. The eqx.Module one (f2 function) is slower (27.8 us on CPU for gradient evaluation) than a trivial implementation (f1; 11.5 us on CPU for gradient evaluation) though the flops of f2 is smaller than that of f1. Is this something I should have expected?

```
import functools as ft
import timeit
import jax
from jax import jit, vmap, lax, jacfwd, jacrev, grad, vjp, jvp, random
import jax.numpy as jnp
import equinox as eqx
import lineax as lx
from jax.config import config
from jaxtyping import Array, Bool, Float, Scalar
from typing import Union
config.update("jax_enable_x64", True)

x_test = jnp.linspace(0, 1, 101)
y_test = jnp.linspace(0, 1, 101)

state_varialbes_test = jnp.hstack((x_test, y_test))

class State(eqx.Module):
    x: Float[Array, "dim1"]
    y: Float[Array, "dim1"]

state = State(x_test, y_test)

def f(state_variables, alpha, beta):
    x = state_variables[0:101]
    y = state_variables[101:]

    z = alpha*x**2 + beta*x*y + y
    
    return jnp.sum(z)

def f2(state, alpha, beta):

    z = alpha*state.x**2 + beta*state.x*state.y + state.y
    
    return jnp.sum(z)

alpha = 2.0
beta = 3.0

jitted_f = jit(f)
jitted_f2 = jit(f2)

jitted_f(state_varialbes_test, alpha, beta)

jitted_f2(state, alpha, beta)

%timeit jax.block_until_ready(jitted_f(state_varialbes_test, alpha, beta)) # 11.4 us on CPU

%timeit jax.block_until_ready(jitted_f2(state, alpha, beta)) # 18.6 us on CPU

grad_f = grad(f)
grad_f2 = grad(f2)

jitted_grad_f = jit(grad_f)
jitted_grad_f2 = jit(grad_f2)

jitted_grad_f(state_varialbes_test, alpha, beta)

jitted_grad_f2(state, alpha, beta)

%timeit jax.block_until_ready(jitted_grad_f(state_varialbes_test, alpha, beta)) # 11.5 us on CPU

%timeit jax.block_until_ready(jitted_grad_f2(state, alpha, beta)) # 27.8 us on CPU

lowered = jitted_grad_f.lower(state_varialbes_test, alpha, beta)
compiled = lowered.compile()
compiled.cost_analysis()[0]['flops'] # 708

lowered2 = jitted_grad_f2.lower(state, alpha, beta)
compiled2 = lowered2.compile()
compiled2.cost_analysis()[0]['flops'] # 506
```

**Comment by cottrell** [comment]
I was curious about this, minor clean up and added another test function

```python
# https://github.com/patrick-kidger/equinox/issues/576
import equinox as eqx
import jax
import jax.numpy as jnp
from jax import grad, jit
from jax.config import config
from jax.lib import xla_bridge
from jaxtyping import Array, Float

platform = 'cpu'
# platform = 'gpu'
jax.config.update('jax_platform_name', platform)
print(xla_bridge.get_backend().platform)

config.update("jax_enable_x64", True)

x_test = jnp.linspace(0, 1, 101)
y_test = jnp.linspace(0, 1, 101)


class State(eqx.Module):
    x: Float[Array, "dim1"]
    y: Float[Array, "dim1"]


state_variables_test = jnp.hstack((x_test, y_test))
state_dict = dict(x=x_test, y=y_test)
state = State(x=x_test, y=y_test)


def f(x, y, alpha, beta):
    z = alpha * x**2 + beta * x * y + y
    return jnp.sum(z)


def f1(state_variables, alpha, beta):
    x = state_variables[0:101]
    y = state_variables[101:]
    return f(x, y, alpha, beta)


def f2(state, alpha, beta):
    x = state.x
    y = state.y
    return f(x, y, alpha, beta)


def f3(state_dict, alpha, beta):
    x = state_dict['x']
    y = state_dict['y']
    return f(x, y, alpha, beta)


alpha = 2.0
beta = 3.0

jitted_f = jit(f)
jitted_f1 = jit(f1)
jitted_f2 = jit(f2)
jitted_f3 = jit(f3)

jitted_f1(state_variables_test, alpha, beta)
jitted_f2(state, alpha, beta)
jitted_f3(state_dict, alpha, beta)

"""
%timeit jax.block_until_ready(jitted_f1(state_variables_test, alpha, beta))
%timeit jax.block_until_ready(jitted_f2(state, alpha, beta))
%timeit jax.block_until_ready(jitted_f3(state_dict, alpha, beta))

gpu
110 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
118 µs ± 1.56 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
114 µs ± 4.98 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

cpu
12.6 µs ± 392 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
19.1 µs ± 623 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
13.4 µs ± 181 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
"""

grad_f1 = grad(f1)
grad_f2 = grad(f2)
grad_f3 = grad(f3)
jitted_grad_f1 = jit(grad_f1)
jitted_grad_f2 = jit(grad_f2)
jitted_grad_f3 = jit(grad_f3)

jitted_grad_f1(state_variables_test, alpha, beta)
jitted_grad_f2(state, alpha, beta)
jitted_grad_f3(state_dict, alpha, beta)

"""
%timeit jax.block_until_ready(jitted_grad_f1(state_variables_test, alpha, beta)) 
%timeit jax.block_until_ready(jitted_grad_f2(state, alpha, beta))
%timeit jax.block_until_ready(jitted_grad_f3(state_dict, alpha, beta))

gpu
105 µs ± 1.63 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
147 µs ± 2.86 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
122 µs ± 3.54 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

cpu
12.7 µs ± 537 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
29.7 µs ± 2.48 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
16.9 µs ± 551 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
"""

"""
lowered = jitted_grad_f1.lower(state_variables_test, alpha, beta)
compiled = lowered.compile()
print(compiled.cost_analysis()[0]['flops'])
lowered2 = jitted_grad_f2.lower(state, alpha, beta)
compiled2 = lowered2.compile()
print(compiled2.cost_analysis()[0]['flops'])
lowered3 = jitted_grad_f3.lower(state_dict, alpha, beta)
compiled3 = lowered3.compile()
print(compiled3.cost_analysis()[0]['flops'])

gpu
708.0
506.0
506.0

cpu
708.0
506.0
506.0
"""
```

**Comment by ToshiyukiBandai** [comment]
Hi David,

Thank you for checking them. I tested your code and got qualitatively similar results. I have another example of lower performance when I used an eqx.Module class as an argument even if the flops number was identical to a trivial implementation. Maybe, it took some time for Python to search class attributes in a class (or keys in a dictionary), and they were not optimized by XLA?

Also, a bit of surprised by the lower performance of GPU. I got a better performance of GPU for a more complicated problems, so it may just depend on the size of the problem.

**Comment by cottrell** [comment]
@ToshiyukiBandai slower gpu perf is typical for small test problems so that is fine. The relative perf of the equinox module is interesting and hopefully we can figure it out.

**Comment by ToshiyukiBandai** [comment]
@cottrell I've tested using __slots__ in the class, but it seems not compatible with JIT... Looking at the fact that the GPU result of the equinox module was not as bad as the CPU one, it may have been caused by how much XLA could optimize the codes. But, in another more complicated case, the equinox on GPU was three times slower than without equinox modules...

@patrick-kidger Do you happen to know what is happening?

**Comment by patrick-kidger** [comment]
I can see there's quite a lot extra going on in these MWE (e.g. multiple test cases, using jaxtyping, etc.) Can you reduce this down to the smallest single example you can, and then I'll be able to take a look at it?

**Comment by ToshiyukiBandai** [comment]
Hi Patrick,

Here is the smallest example I could make. Thank you for your help!

```
import functools as ft
import timeit
import jax
from jax import jit, grad
import jax.numpy as jnp
import equinox as eqx
from jax.config import config

config.update("jax_enable_x64", True)

class State(eqx.Module):
    x: jax.Array

def f(x, alpha):
    z = alpha*x**2
    return jnp.sum(z)

def f2(state, alpha):
    z = alpha*state.x**2
    return jnp.sum(z)

alpha = 2.0
x_test = jnp.linspace(0, 1, 101)
state = State(x_test)

jitted_f = jit(f)
jitted_f2 = jit(f2)

jitted_f(x_test, alpha)
jitted_f2(state, alpha)

%timeit jax.block_until_ready(jitted_f(x_test, alpha))  # 10.1 us on CPU
%timeit jax.block_until_ready(jitted_f2(state, alpha))  # 16.3 us on CPU

grad_f = grad(f)
grad_f2 = grad(f2)

jitted_grad_f = jit(grad_f)
jitted_grad_f2 = jit(grad_f2)

jitted_grad_f(x_test, alpha)
jitted_grad_f2(state, alpha)

%timeit jax.block_until_ready(jitted_grad_f(x_test, alpha))  # 9.68 us on CPU
%timeit jax.block_until_ready(jitted_grad_f2(state, alpha))  # 22.4 us on CPU
```

**Comment by patrick-kidger** [comment]
Thanks!
Looking at your timings I think this is expected: flattening the Module down to its constituent arrays takes a small amount of time -- in this case, a few microseconds.

I think that's a small enough number that this overhead is acceptable. Doing the actual work inside most JIT'd computations tends to be substantially longer than this.

If you have a particularly short computation, comparable to the very simple examples you have here, then you could either consider packing multiple computations together into a `jax.lax.scan` (with a single JIT wrapper around that), or where applicable try the telescoping trick [described here](https://docs.kidger.site/equinox/tricks/#low-overhead-training-loops).

**Comment by ToshiyukiBandai** [comment]
Thank you! That makes sense. I will try the telescoping trick too!
